tasks/FileUploadTask.md

```diff
@@ -9,26 +9,29 @@ This task aims to provide a fluent and easy to use mechanism for the consumer to
 
 - Accept a Microsoft.Graph.UploadSession instance to be used for the upload which is created and managed by the user and handed over to the task. The task should use the instance to start/resume the upload.
 - The task should validate the expiration time and expected byte range of the upload session on resume by querying the upload URL(thereby updating its instance of the Microsoft.Graph.UploadSession).
-- Accept a file/stream to be uploaded
-- The task should provide a mechanism of configuring the upload properties i.e. the upload slice size
-- The task should provide a mechanism of monitoring the upload status i.e the progress and failures through a callback/Delegate
-- Using the RequestContext, the feature flag for the FileUploadTask can be set for telemetry purposes.
-- In the event that a slice fails to upload, the task should retry to re-upload the slice for a configurable number of times(default of 3) before giving up and throwing an error/exception for the user to handle.
-- The task should be agnostic to the kind of upload being perfromed so as to support for various fileUpload scenarios e.g. DriveItem and FileAttachment
+- Accept a file/stream to be uploaded.
+- The task should provide a mechanism of configuring the upload properties i.e. the upload slice size.
+- The task should provide a mechanism of monitoring the upload progress through a callback/delegate.
+- The task should signal upload failures via standard exception mechanism.
+- The task should signal upload completion via by returning the status or returning a completed **Task**/**Promise**/**Future** for async APIs.
```

Comment:
We should elaborate more on how to process the completion response. 
1. Use an `UploadResult` object and its design.
2. Processing the response and what to return.  State that different APIs have different responses. 
This will help in consistency with the SDKs. 
@andrueastman What do you think?

Reply:
I would vouch for option 1. 
Trying to process the result(e.g by calling the url in the location header) could result in unintended issues/errors/complexity. 
The location url could sometimes appear as a none graph url(having an outlook base url) therefore meaning our call could potentially not have the right scopes or auth token(especially since the client was just configured to not set the auth header for uploading).
I think it would make more sense to give the information to the user(the id/location header) in the UploadResult object and let them process it as they please.

Reply:
Following this logic, our UploadResult would look something like that (this is pseudo json schema), this type would be generic so the object prop would be strongly typed)
```json
{
  "type": "object",
  "properties": {
    "location": { "type": "string"},
    "id": { "type": "string"},
    "object": { "type":  "any"},
  }
}
```

If we get a location response header, we populate the location field here, if we get an id, populate the id, if we get an object, populate the object. What do you think?

Currently in Java, if we get an id, we'll populate the object type (DriveItem or whatever) with just the Id, instead of returning an Id flat out. Do you think it's better? Worse? (I believe it simplifies things for consumer as they have one less path to care about).

Reply:
Is there a scenario where we would only get an Id without an object?
Otherwise, I agree that it would make a lot of sense to have one less path to think about as it is done in Java.

Reply:
That's not up to the consumer, it's up to the service implementing large upload. So besides that scenario where we only get the id after the upload is completed, I can't think of any existing or potential scenario why we'd get only the id.

Reply:
I recommend the body as is, return it to the user and leave it to the user on how the id should be processed.
Because outlook does not return a body or an ID. 
Any change in the future to the body structure or say id property should not block the processing. 
As as far I know, the Location property is documented to be in the header for all 3- Onedrive, Outlook, PrintDocument APIs. 
I would assume it is safe to read it from the header. However, I am concerned about the case of the Location header key.

Reply:
yep, I think we're on the same page. That's what I updated it to besides the name of the property which I'll update. Resolving.